### PR TITLE
Raise a Liquid::ArgumentError in slice filter for invalid integers.

### PR DESCRIPTION
**Pull request by dylanahsmith**
@tjoyal & @fw42 for review

If an invalid integer is passed to slice for the offset of length arguments, then raise a Liquid::ArgumentError instead of a non-Liquid::Error, since this is a template bug.

### CODE REVIEW

**Comment by tjoyal** [comment]
What was the original error looking like? Could this new custom error be abused into generating a new xss exploit point?

**Comment by dylanahsmith** [comment]
> What was the original error looking like? Could this new custom error be abused into generating a new xss exploit point?

I was seeing the error `can't convert nil into Integer`

If the input were a string, then it would also include the input

```
irb> Integer('<script>alert(1)</script>')
ArgumentError: invalid value for Integer(): "<script>alert(1)</script>"
```

but that would only be exploitable if the visitor had control of the input.

**Comment by tjoyal** [comment]
:+1: then as it was already exploitable

File: lib/liquid/standardfilters.rb [hunk 1]
```diff
@@ -317,6 +317,16 @@ def default(input, default_value = "".freeze)
 
     private
 
+    def to_integer(num)
+      return num if num.is_a?(Integer)
+      num = num.to_s
+      begin
+        Integer(num)
```
Comment (fw42) [inline lib/liquid/standardfilters.rb]: can you add a test for this line? i.e. one where the argument is a string that can be converted to an integer?

**Comment by fw42** [comment]
:+1:

File: lib/liquid/standardfilters.rb [hunk 2]
```diff
@@ -317,6 +317,16 @@ def default(input, default_value = "".freeze)
...
+      rescue ::ArgumentError
+        raise Liquid::ArgumentError, "invalid integer '#{num}'"
```
Comment (fw42) [inline lib/liquid/standardfilters.rb]: do we really need to put the `#{num}` in here? does that give us much? I'm concerned now about the possible XSS you mentioned.

Comment (dylanahsmith) [inline lib/liquid/standardfilters.rb]: it can help debugging, but I guess it can be removed

File: lib/liquid/standardfilters.rb (hunk 1 above)
Comment (dylanahsmith) [inline lib/liquid/standardfilters.rb]: done

File: lib/liquid/standardfilters.rb (hunk 2 above)
Comment (dylanahsmith) [inline lib/liquid/standardfilters.rb]: removed ` '#{num}'` from the error message

**Pull request merged**